apps/spack/azure_blob.py

```diff
@@ -0,0 +1,100 @@
+
+"""This file contains the definition of the Azure Blob storage Class used to integrate Azure Blob storage with spack buildcache.
+"""
+
+import os
+import sys
+import datetime
+from pathlib import Path
+
+import llnl.util.tty as tty
+
+class AzureBlob:
+    def __init__(self, url):
+        from azure.storage.blob import BlobServiceClient
+        self.url = url
+        (self.container_name, self.blob_path) = self.get_container_blob_path()
+        if url.scheme != 'azure':
+           raise ValueError(
+            'Can not create Azure blob connection from URL with scheme: {SCHEME}'.format(
+                SCHEME=url.scheme))
+        if "AZURE_STORAGE_CONNECTION_STRING" in os.environ:
+           self.connect_str = os.environ.get('AZURE_STORAGE_CONNECTION_STRING')
+           self.blob_service_client = BlobServiceClient.from_connection_string(self.connect_str)
+           if not self.azure_container_exists():
+              tty.warn("The container {} does not exist, it will be created".format(self.container_name))
+              self.blob_service_client.create_container(self.container_name)
+        else:
+           tty.error("Error: Environmental variable AZURE_STORAGE_CONNECTION_STRING is not defined, it is required if you want to use Azure Blob storage as an spack buildcache")
+           sys.exit(1)
+
+    def get_container_blob_path(self):
+        blob_path = self.url.path
+        p = Path(blob_path)
+        container_name = p.parts[1]
+        blob_path = str(Path(*p.parts[2:]))
+        tty.debug("container_name = {}, blob_path = {}".format(container_name, blob_path))
+        return (container_name, blob_path)
+
+
+    def is_https(self):
+        if 'https' in self.connect_str:
+           return True
+        else:
+           return False
+
+
+    def azure_container_exists(self):
+        try:
+           ttt = self.blob_service_client.get_container_client(self.container_name).list_blobs()
+           for blob in ttt:
+               pass
+        except Exception as ex:
+            return False
+        return True
+
+
+    def azure_blob_exists(self):
+        try:
+           blob_client = self.blob_service_client.get_blob_client(container=self.container_name, blob=self.blob_path)
+           blob_properties = blob_client.get_blob_properties()
+        except Exception as ex:
+            return False
+        return True
+
+
+    def azure_upload_to_blob(self, local_file_path):
+        from azure.storage.blob import ContentSettings
+        if Path(self.blob_path).suffix == '.json':
+           contentsettings = ContentSettings(content_type="application/json")
+        else:
+           contentsettings = ContentSettings()
+        blob_client = self.blob_service_client.get_blob_client(container=self.container_name, blob=self.blob_path)
+        with open(local_file_path, "rb") as data:
+            blob_client.upload_blob(data, overwrite=True, content_settings=contentsettings)
```

Comment:
Is there any retry logic here ? if not then there should be one. 
how does errors are handled and recovered ?

Reply:
I am using an Azure Blob Storage python SDK, I would expect that it would handle some of the lower level re-try login you would need in RESP API? I have added some logic to catcc and report any failures.

Reply:
Retry policy is on by default see there https://docs.microsoft.com/en-us/azure/developer/python/sdk/storage/storage-blob-readme?view=storage-py-v12